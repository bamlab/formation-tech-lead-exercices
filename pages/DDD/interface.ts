import {PriceInCents} from './priceUtils';

export interface Product {
  id: number;
  price: PriceInCents;
  name: string;
}

export interface Discount {
  id: number;
  value: number;
  type: string;
}
