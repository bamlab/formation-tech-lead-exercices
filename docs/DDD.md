# DDD

## What we are trying to do

We display a list of products with their price coming from a fake API. There can be a discount applied to the products.

There is a button at the bottom displaying the total amount of the basket.

![](./basket.png)

## Problems

### Applying a discount does not display a round price

Even if we manipulate price in cents to avoid rounding issues (such as `0.2 + 0.1 = 0.30000000002`), we forgot to be cautious about the discount!
Make use of TS brands to prevent from displaying a number that is not a price in cents.

### Actually, a discount can be applied to one product only

Ah! That wasn't in the initial requirement, but when testing the app the product team realised the case where a discount is applied to 1 product only is not applied!
In this case, the discount must be applied directly in the displayed price - it should be transparent to the user.

To test this case, change the basketId to `2`.

Make use of adapters to make this change safely :)

### Aaaargh, a mega Beatles fan finds their app really slow

There is a super Beatles fan that purchased 2000 Beatles albums.
However, when they try to order their products by price, the app is really slow.

To test this case, change the basketId to `3`.

Normalise the data to make it faster :)
