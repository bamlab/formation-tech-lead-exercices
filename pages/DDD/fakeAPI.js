const createBaseBasket = () => ({
  products: [
    {
      id: 1,
      name: 'Post its bleus',
      price: 12.39,
    },
    {
      id: 2,
      name: 'Post its rouges',
      price: 17.47,
    },
  ],
  discounts: [
    {
      id: 1,
      type: 'percent',
      value: 10,
    },
  ],
});

export const getProducts = async basketId => {
  if (basketId === undefined) throw new Error('no basket id');
  // fake timeout
  await new Promise(resolve => {
    setTimeout(() => resolve(), 200);
  });

  const basket = createBaseBasket();

  // basket with discounts on product
  if (basketId === 2) {
    return {
      ...basket,
      discounts: [
        {
          id: 2,
          type: 'percent',
          value: 20,
          productId: 1,
        },
      ],
    };
  }

  // basket with a lot of products
  if (basketId === 3) {
    return {
      ...basket,
      products: Array.from(Array(200)).map((_, index) => ({
        id: index,
        name: `album des Beatles #${index}`,
        price: Math.round(Math.random() * 1000),
      })),
    };
  }

  return basket;
};
