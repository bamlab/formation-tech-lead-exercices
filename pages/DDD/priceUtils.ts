import {Discount, Product} from './interface';

export type PriceInCents = number & {_brand: 'PriceInCents'};

export const computeProductsTotalPrice = (
  products: Product[],
): PriceInCents => {
  return products.reduce(
    (total, product) => total + product.price,
    0,
  ) as PriceInCents;
};

export const applyDiscountToPrice = (
  discounts: Discount[],
  priceWithoutDiscounts: PriceInCents,
): PriceInCents => {
  const price = discounts.reduce<number>((total, discount) => {
    return (total * (100 - discount.value)) / 100;
  }, priceWithoutDiscounts);
  return Math.floor(price) as PriceInCents;
};

export const formatPriceInCents = (price: PriceInCents): string =>
  `${price / 100}`;
