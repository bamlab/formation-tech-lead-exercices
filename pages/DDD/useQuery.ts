import {useEffect, useState} from 'react';

export function useQuery<Data>(query: () => Promise<Data>) {
  const [data, setData] = useState<Data | null>(null);
  const [loading, setLoading] = useState(true);

  const loadData = async () => {
    const d = await query();
    setData(d);
    setLoading(false);
  };

  useEffect(() => {
    loadData();
  }, []);

  return {data, loading};
}
