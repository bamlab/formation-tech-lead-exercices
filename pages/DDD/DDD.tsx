/* eslint-disable react-native/no-inline-styles */
import React, {useEffect, useState} from 'react';
import {View, Text, FlatList, Button, Alert} from 'react-native';
import {Discount, Product} from './interface';
//@ts-ignore
import {getProducts} from './fakeAPI';

const ProductItem = (props: {product: Product}) => {
  return (
    <View>
      <Text>
        {props.product.name} - {props.product.price}€
      </Text>
    </View>
  );
};

const Header = () => {
  return <Text>Products</Text>;
};

const PaymentButton = (props: {
  data: {products: Product[]; discounts: Discount[]};
}) => {
  const price = props.data.products.reduce(
    (total, product) => total + product.price,
    0,
  );

  return (
    <Button
      title={`Payer ${price}€`}
      onPress={() => Alert.alert(`${price}€`)}
    />
  );
};

function useQuery<Data>(query: () => Promise<Data>) {
  const [data, setData] = useState<Data | null>(null);
  const [loading, setLoading] = useState(true);

  const loadData = async () => {
    const d = await query();
    setData(d);
    setLoading(false);
  };

  useEffect(() => {
    loadData();
  }, []);

  return {data, loading};
}

export const DDD = () => {
  const {data, loading} = useQuery<any>(() => getProducts(1));

  return (
    <View>
      {loading ? (
        <Text>Loading</Text>
      ) : (
        <FlatList
          ListHeaderComponent={<Header />}
          data={data.products}
          renderItem={info => <ProductItem product={info.item} />}
          ListFooterComponent={<PaymentButton data={data} />}
        />
      )}
    </View>
  );
};
