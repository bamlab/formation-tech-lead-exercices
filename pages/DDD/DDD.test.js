import React from 'react';

import {fireEvent, render, waitFor} from '@testing-library/react-native';
import {DDD} from './DDD';

it('renders correctly', async () => {
  const {toJSON, getByText} = render(<DDD />);

  await waitFor(() => getByText('Products'));
  expect(toJSON()).toMatchSnapshot();

  // Check the order is reversed with the price
  const orderByPriceButton = getByText('tri par prix');
  fireEvent.press(orderByPriceButton);

  expect(toJSON()).toMatchSnapshot();
});
