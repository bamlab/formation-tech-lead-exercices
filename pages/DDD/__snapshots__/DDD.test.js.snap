// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <RCTScrollView
    ListFooterComponent={
      <PaymentButton
        data={
          Object {
            "discount": Object {
              "id": 1,
              "type": "percent",
              "value": 10,
            },
            "products": Array [
              Object {
                "id": 1,
                "name": "Post its bleus",
                "price": 12.39,
              },
              Object {
                "id": 2,
                "name": "Post its rouges",
                "price": 17.47,
              },
            ],
          }
        }
      />
    }
    ListHeaderComponent={<Header />}
    data={
      Array [
        Object {
          "id": 1,
          "name": "Post its bleus",
          "price": 12.39,
        },
        Object {
          "id": 2,
          "name": "Post its rouges",
          "price": 17.47,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <Text>
          Products
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text>
            Post its bleus
             - 
            12.39
            €
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text>
            Post its rouges
             - 
            17.47
            €
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityRole="button"
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Payer 29.86€
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
